src/osp/id_registry.h

```diff
@@ -0,0 +1,341 @@
+
+namespace osp
+{
+
+template<class TYPE_T>
+constexpr TYPE_T id_null() noexcept
+{
+    using underlying_t = typename std::underlying_type_t<TYPE_T>;
+    return TYPE_T(std::numeric_limits<underlying_t>::max());
+}
+
+//-----------------------------------------------------------------------------
+
+/**
+ * @brief Generates reusable sequential IDs
+ */
+template<typename ID_T, bool NO_AUTO_RESIZE = false>
+class IdRegistry
+{
+    using id_int_t = std::underlying_type_t<ID_T>;
+
+public:
+
+    IdRegistry() = default;
+    IdRegistry(size_t capacity) { reserve(capacity); };
+
+    /**
+     * @brief Create a single ID
+     *
+     * @return A newly created ID
+     */
+    [[nodiscard]] ID_T create()
+    {
+        ID_T output;
+        create(&output, 1);
+        return output;
+    }
+
+    /**
+     * @brief Create multiple IDs
+     *
+     * @param out   [out] Iterator out
+     * @param count [in] Number of IDs to generate
+     */
+    template<typename IT_T>
+    void create(IT_T out, size_t count);
+
+    /**
+     * @return Size required to fit all existing IDs, or allocated size if
+     *         reserved ahead of time
+     */
+    constexpr size_t capacity() const noexcept { return m_deleted.size(); }
+
+    /**
+     * @return Number of registered IDs
+     */
+    constexpr size_t size() const { return capacity() - m_deleted.count(); }
+
+    /**
+     * @brief Allocate space for at least n IDs
+     *
+     * @param n [in] Number of IDs to allocate for
+     */
+    void reserve(size_t n)
+    {
+        m_deleted.resize(n, true);
+    }
+
+    /**
+     * @brief Remove an ID. This will mark it for reuse
+     *
+     * @param id [in] ID to remove
+     */
+    void remove(ID_T id)
+    {
+        if ( ! exists(id))
+        {
+            throw std::runtime_error("ID does not exist");
+        }
+
+        m_deleted.set(id_int_t(id));
+    }
+
+    /**
+     * @brief Check if an ID exists
+     *
+     * @param id [in] ID to check
+     *
+     * @return true if ID exists
+     */
+    bool exists(ID_T id) const noexcept
+    {
+        return ! m_deleted.test(id_int_t(id));
+    };
+
+    template <typename FUNC_T>
+    void for_each(FUNC_T func);
+
+private:
+    HierarchicalBitset<uint64_t> m_deleted;
+
+}; // class IdRegistry
+
+template<typename ID_T, bool NO_AUTO_RESIZE>
+template<typename IT_T>
+void IdRegistry<ID_T, NO_AUTO_RESIZE>::create(IT_T out, size_t count)
+{
+    if (m_deleted.count() < count)
+    {
+        // auto-resize
+        if constexpr (NO_AUTO_RESIZE)
+        {
+            throw std::runtime_error(
+                    "Reached max capacity with automatic resizing disabled");
+        }
+        else
+        {
+            reserve(std::max(capacity() + count, capacity() * 2));
+        }
+    }
+
+    m_deleted.take<IT_T, ID_T>(out, count);
+}
+
+template<typename ID_T, bool NO_AUTO_RESIZE>
+template <typename FUNC_T>
+void IdRegistry<ID_T, NO_AUTO_RESIZE>::for_each(FUNC_T func)
```

Comment:
if you decide not to remove this and replace it with begin() and end() functions, then i recommend taking FUNC_T as && to avoid always copying the provided function.